drivers/net/w5500/w5500.c

```diff
@@ -0,0 +1,1099 @@
+{
+	int ret;
+	uint8_t status;
+
+	if (sock_id > W5500_MAX_SOCK_NUMBER)
+		return -EINVAL;
+
+	ret = w5500_socket_reg_read(dev, sock_id, W5500_Sn_SR, &status, 1);
+	if (ret)
+		return ret;
+
+	if (status != W5500_Sn_SR_ESTABLISHED &&
+	    status != W5500_Sn_SR_CLOSE_WAIT)
+		return -ENOTCONN;
+
+	return w5500_socket_command_write(dev, sock_id, W5500_Sn_CR_DISCON);
+}
+
+/***************************************************************************//**
+ * @brief Send data through a socket
+ *
+ * @param dev     - Device descriptor
+ * @param sock_id - Socket number (0-7)
+ * @param buf     - Data buffer to send
+ * @param len     - Data length
+ *
+ * @return Number of bytes sent on success, negative error code otherwise
+*******************************************************************************/
+int w5500_socket_send(struct w5500_dev *dev, uint8_t sock_id, const void *buf,
+		      uint16_t len)
+{
+	int ret;
+	uint16_t free_size = 0;
+	uint16_t wr_ptr;
+	uint8_t wr_ptr_buf[2];
+	uint8_t status;
+	uint16_t total_sent = 0;
+	uint16_t chunk_size;
+	const uint8_t *data = (const uint8_t *)buf;
+
+	if (sock_id > W5500_MAX_SOCK_NUMBER)
+		return -EINVAL;
+
+	ret = w5500_socket_reg_read(dev, sock_id, W5500_Sn_SR, &status, 1);
+	if (ret)
+		return ret;
+
+	if (status != W5500_Sn_SR_ESTABLISHED &&
+	    status != W5500_Sn_SR_UDP &&
+	    status != W5500_Sn_SR_MACRAW)
+		return -ENOTCONN;
+
+	while (total_sent < len) {
+		ret = w5500_read_16bit_reg(dev, W5500_SOCKET_REG_BLOCK(sock_id),
+					   W5500_Sn_TX_FSR, &free_size);
+		if (ret)
+			return ret;
+
+		if (free_size == 0) {
+			no_os_mdelay(1);
+			continue;
+		}
+
+		chunk_size = (len - total_sent) < free_size ? (len - total_sent) : free_size;
+
+		ret = w5500_read_16bit_reg(dev, W5500_SOCKET_REG_BLOCK(sock_id),
+					   W5500_Sn_TX_WR, &wr_ptr);
+		if (ret)
+			return ret;
+
+		ret = w5500_reg_write(dev, W5500_SOCKET_TX_BUF_BLOCK(sock_id),
+				      wr_ptr, data + total_sent, chunk_size);
+		if (ret)
+			return ret;
+
+		wr_ptr += chunk_size;
+
+		wr_ptr_buf[0] = W5500_BYTE_HIGH(wr_ptr);
+		wr_ptr_buf[1] = W5500_BYTE_LOW(wr_ptr);
+
+		ret = w5500_socket_reg_write(dev, sock_id, W5500_Sn_TX_WR,
+					     wr_ptr_buf, 2);
+		if (ret)
+			return ret;
+
+		ret = w5500_socket_command_write(dev, sock_id, W5500_Sn_CR_SEND);
+		if (ret)
+			return ret;
+
+		total_sent += chunk_size;
+	}
+
+	return len;
+}
+
+/***************************************************************************//**
+ * @brief Receive data from a socket
+ *
+ * @param dev     - Device descriptor
+ * @param sock_id - Socket number (0-7)
+ * @param buf     - Buffer to store received data
+ * @param len     - Maximum length to receive
+ *
+ * @return Number of bytes received on success, negative error code otherwise
+*******************************************************************************/
+int w5500_socket_recv(struct w5500_dev *dev, uint8_t sock_id, void *buf,
```

Comment:
By default, `recv()` is a blocking function. I think you should implement it the same way you did with `send()` above.